### PR TITLE
#615, Support other PHP executable

### PR DESCRIPTION
**Pull request by rotemreiss**
| Q             | A
| ------------- | ---
| Branch        | master for features and deprecations
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Documented?   | yes
| Fixed tickets | #615 

See #615 - support running grumphp on multiple locally installed PHP versions.

### CODE REVIEW

**Comment by veewee** [comment]
Hello @rotemreiss,

The PR looks good, thanks! I think you missed one reference in the general README file.
I'll have to make time to manually test this PR before I can merge it in.

**Comment by rotemreiss** [comment]
> Hello @rotemreiss,
> 
> The PR looks good, thanks! I think you missed one reference in the general README file.
> I'll have to make time to manually test this PR before I can merge it in.

Added the missing README.
Thanks.

**Comment by jeckel** [comment]
I like the idea, but I think there is something wrong in this PR.

When using the `exec` command, parameters are quoted like `'--ansi' --git-user='$GIT_USER' --git-email='$GIT_EMAIL'...`

But if you use something else (like docker), quoting paremeters might not works.

In the other hand, if you want to need a multipart executable (with additional parameters) both this solution didn't worked :

```yaml
EXEC_GRUMPHP_COMMAND: "docker-compose run --rm --no-deps php"
```
And this one is not valid yaml:
```yaml
EXEC_GRUMPHP_COMMAND: "docker-compose" "run" "--rm" "--no-deps" "php"
```

**Comment by ryantfowler** [comment]
@veewee any idea when this may get merged in? In the meantime, if others are interested in making this work outside of the scope of this PR, here's what I've got working:
### pre-commit
```
#!/bin/sh

# Run the hook command.
# Note: this will be replaced by the real command during copy.

# Fetch the GIT diff and format it as command input:
DIFF=$(git -c diff.mnemonicprefix=false --no-pager diff -r -p -m -M --full-index --no-color --staged | cat)

# Check if we're on a host system or within a Docker container
if ! [ -x "$(command -v php)" ];then
      (cd "./" && echo "${DIFF}" | \
            exec 'docker' 'run' '--rm' '-t' "--volume=$PWD:{{{directory_path_in_container}}}" '{{{docker_image}}}' \
            'vendor/bin/grumphp' 'git:pre-commit' '--skip-success-output')
else
    (cd "./" && echo "${DIFF}" | exec 'vendor/bin/grumphp' 'git:pre-commit' '--skip-success-output')
fi
```

### commit-msg
```
#!/bin/sh

#
# Run the hook command.
# Note: this will be replaced by the real command during copy.
#

GIT_USER=$(git config user.name)
GIT_EMAIL=$(git config user.email)
COMMIT_MSG_FILE=$1

# Fetch the GIT diff and format it as command input:
DIFF=$(git -c diff.mnemonicprefix=false --no-pager diff -r -p -m -M --full-index --no-color --staged | cat)

# Check if we're on a host system or within a Docker container
if ! [ -x "$(command -v php)" ];then

    (cd "./" && echo "${DIFF}" | \
            exec 'docker' 'run' '--rm' '-t' "--volume=$PWD:{{{directory_path_in_container}}}" '{{{docker_image}}}' \
            'vendor/bin/grumphp' 'git:pre-commit' "--git-user=$GIT_USER" "--git-email=$GIT_EMAIL" "$COMMIT_MSG_FILE")
else
    (cd "./" && printf "%s\n" "${DIFF}" | exec '/var/www/magento/vendor/bin/grumphp' 'git:commit-msg' "--git-user=$GIT_USER" "--git-email=$GIT_EMAIL" "$COMMIT_MSG_FILE")
fi
```

Where `{{{directory_path_in_container}}}` and `{{{docker_image}}}` should be switched out with the actual values needed for a given project

**Comment by veewee** [comment]
@ryantfowler Thanks for sharing!
I am quite busy at the moment so it will take me some time before I can test this.
Maybe @Landerstraeten  can pick it up?

**Comment by veewee** [comment]
Hello @rotemreiss,

I've tested the PR and it works fine for simple executables. For more complex executables, I changed some lines in the init command. Those should work as well now.
Don't have docker running anymore. It would be great if @jeckel could review the changes as well...

@Landerstraeten : can you also take a second look at this PR?

**Comment by veewee** [comment]
(Not sure why appveyor fails, gonna have to look at it another time...)

**Comment by jeckel** [comment]
I'm testing this today with docker and docker-compose

File: resources/hooks/local/commit-msg [hunk 1]
```diff
@@ -13,4 +13,4 @@ COMMIT_MSG_FILE=$1
 DIFF=$(git -c diff.mnemonicprefix=false --no-pager diff -r -p -m -M --full-index --no-color --staged | cat)
 
 # Run GrumPHP
-(cd "${HOOK_EXEC_PATH}" && printf "%s\n" "${DIFF}" | exec $(HOOK_COMMAND) "--git-user=$GIT_USER" "--git-email=$GIT_EMAIL" "$COMMIT_MSG_FILE")
+(cd "${HOOK_EXEC_PATH}" && printf "%s\n" "${DIFF}" | $(EXEC_GRUMPHP_COMMAND) $(HOOK_COMMAND) "--git-user=$GIT_USER" "--git-email=$GIT_EMAIL" "$COMMIT_MSG_FILE")
```
Comment (jeckel) [inline resources/hooks/local/commit-msg]: There is an error if $GIT_USER contains spaces like 'firstname lastname'.

It should work with `"--git-user='$GIT_USER'"`

Comment (veewee) [reply resources/hooks/local/commit-msg]: strange, the argument is allready wrapped with quotes ... Will investigate it. Thanks for testing!
Does it work on docker + docker-compose with a single worded username?

**Comment by rotemreiss** [comment]
Tested it with `EXEC_GRUMPHP_COMMAND: "lando php"` and it works like a charm. :)

@veewee Thanks for the great effort in this!

File: resources/hooks/local/commit-msg (hunk 1 above)
Comment (jeckel) [reply resources/hooks/local/commit-msg]: It worked with docker compose and a single worded username

**Comment by ryantfowler** [comment]
@veewee Just curious if this was going to get merged in? Also, any idea when there will be a new release? Thanks

**Comment by veewee** [comment]
Hi @ryantfowler,

Busy weeks ahead... We are planning to put some extra effort into the release at the end of the month. Stay tuned :)

**Comment by veewee** [comment]
@Landerstraeten 
Looks like appveyor can finally handle my code.. :)
Can you have a look at the code + git user issue so that we can ship it?

**Comment by Landerstraeten** [comment]
@veewee the code like fine to me. I haven't test it yet. I will merge it as soon as possible.

**Comment by Landerstraeten** [comment]
Tested with docker `docker run --rm -it -v $(pwd):/grumphp -w /grumphp webdevops/php:alpine bin/grumphp run`
Works like a charm!

**Comment by Landerstraeten** [comment]
Thanks everyone!

**Comment by rotemreiss** [comment]
Thanks! Great work :)

**Pull request merged**